로또 1단계 미션 제출합니다.

chanhalee opened the pull request:
잘 부탁드립니다.

findstar commented on src/main/java/lotto/view/LottoGameView.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Scanner;
+
+public class LottoGameView {
```
사용자 입력을 위한 InputView 와 출력을 위한 OutputView 로 분리하면 좀 더 작은 단위로 클래스를 설계할 수 있습니다.

findstar commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -0,0 +1,41 @@
+        return (int) numbers.stream().filter(comparedTicket::contains).count();
+    }
+
+    public LottoRank getRank(WinnerNumber winnerNumber) {
```
LottoTicket 에서 Rank 를 구하는것과 WinnerNumber 에서 Rank 를 구하는 것 어느것이 더 
역할 분리가 잘 된것일까요?

findstar commented on src/main/java/lotto/domain/RankTable.java:
```diff
@@ -0,0 +1,25 @@
+
+public class RankTable {
+
+    private static final MatchResult FIRST = new MatchResult(6, false);
```
RankTable 을 별도로 분리하기 보다 LottoRank 에서 static method 를 추가하는 것이 어떨까요?

tjrgus1213 replied:
LottoRank는 enum인데 static method나 변수를 가져도 상관없을까요?

findstar replied:
enum 도 static method 를 가질 수 있습니다. 
제가 말씀드린건 static method 를 사용하여 
매칭 카운트와 보너스 정보를 전달하여 
Rank 인스턴스을 생성할 수 있도록 개선해보는 방법이 있다는 것을 알려드린겁니다.

findstar commented on src/main/java/lotto/domain/WinnerNumber.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class WinnerNumber extends LottoTicket {
```
WinnerNumber 와 LottoTicket 상속구조로 연결되면서
WinnerNumber 에는 contains와 같은 메서드 호출이 가능해집니다.
이는 의도한게 아니기 때문에 이 경우에는 상속보다는 분리된 구조의 클래스로 유지하는게 나아보입니다.

tjrgus1213 replied:
LottoTicket과 WinnerNumber 클래스가 공통되는 부분이 많다고 생각하는데 그러면 WinnerNumber가 LottoTicket을 상속받는 것이 아닌 LottoTicket과 WinnerNumber가 같은 클래스를 상속받도록 수정하는건 어떨까요??

findstar replied:
상속은 아주 강한 결합입니다. 
따라서 상위 클래스의 위치에 하위 클래스가 전달되어도 문제없이 사용될 수 있어야 합니다. 
대신 조합(Composition)을 사용해보는건 어떨까요?

findstar commented on src/main/java/lotto/domain/Price.java:
```diff
@@ -0,0 +1,13 @@
+package lotto.domain;
+
+public class Price {
```
아주 간단한 클래스라도 테스트 해주세요 😄

findstar commented on src/main/java/lotto/util/StringUtility.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class StringUtility {
```
유틸리티 클래스를 만들 때는 범용성을 생각하면서 만들어야 합니다. 

다음의 메서드는 모두 도메인 객체 생성을 위한 요구사항의 일부로 볼 수 있네요.

findstar commented on src/main/java/lotto/domain/LottoResults.java:
```diff
@@ -0,0 +1,65 @@
+        return results.get(rank);
+    }
+
+    public double calculateRate(int price) {
```
이런 경우에는 BigDecimal 를 사용하면 수식이 좀 더 간결해집니다.

findstar commented on src/main/java/lotto/view/LottoGameView.java:
```diff
@@ -0,0 +1,43 @@
+        System.out.println("당첨 통계");
+        System.out.println("--------------");
+        System.out.println(lottoResults);
+        System.out.println("총 수익률은 " + lottoResults.calculateRate(price) + " 입니다.");
```
System.out.printf 를 활용해보세요.

findstar commented on README.md:
```diff
@@ -7,3 +7,23 @@
 ## 온라인 코드 리뷰 과정
 * [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
+
+## 구현 클래스
```
클래스 별 역할을 아주 잘 정리 해주셨네요 👍
1단계 요구사항도 추가 부탁드립니다~!

findstar commented on src/main/java/lotto/domain/LottoResults.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public void upsert(LottoRank rank) {
+        if(isNothing(rank))
```
Rank 가 null 인경우가 있을까요? Rank.NOTHING (꽝) 을 추가해보세요.

findstar reviewed (changes requested):
안녕하세요 로또 미션 리뷰를 하게된 제임스입니다.
주말이라 조금 늦게 리뷰를 진행했네요. 수고 많으셨습니다. 👏
다음은 리뷰에 남겨주신 질문에 대한 답변입니다.
main 함수는 말 그대로 애플리케이션을 실행시키는 함수이기 때문에 MVC 와는 구분됩니다.
컨트롤러는 제어의 역할을 수행하는 객체로 이해하시면 됩니다. 
그리고 테스트 코드는 좀 더 유연하게 작성할 수 있습니다. 다만 개별 클래스의 테스트가 일부 빠진 경우가 있으니 보완 부탁드립니다.

몇가지 추가 코멘트를 남겼습니다. 
그리고 코드 커버리지를 확인해보세요. 다음과 같이 인텔리J 에서 커버리지를 확인할 수 있습니다. 

![image](https://user-images.githubusercontent.com/1266944/104094527-bcdbd500-52d4-11eb-948a-1542f828fb6e.png)

그럼 화이팅입니다.

tjrgus1213 commented:
테스트 코드를 좀 더 유연하게 작성해보려고 노력했는데 쉽지 않은 것 같습니다..
어떻게 하면 조금 더 테스트 코드를 유연하게 작성 할 수 있을까요??

findstar commented on src/main/java/lotto/domain/Number.java:
```diff
@@ -0,0 +1,61 @@
+
+    static {
+        numberPool = new ArrayList<>();
+        IntStream.rangeClosed(MIN_LOTTO_NUMBER, MAX_LOTTO_NUMBER).forEach(number -> numberPool.add(new Number(number)));
```
IntStream 을 활용해서 캐싱 하는군요 좋습니다 💯

findstar commented on src/main/java/lotto/domain/WinnerNumber.java:
```diff
@@ -0,0 +1,48 @@
+    private final LottoTicket winnerNumbers;
+    private final Number bonusNumber;
+
+    public WinnerNumber(LottoTicket winnerNumbers, Number bonusNumber) {
```
WinnerNumber 가 상속 구조를 벗어나 조합하는 구조로 변경되었네요 💯

findstar commented on src/main/java/lotto/domain/LottoTicket.java:
```diff
@@ -0,0 +1,70 @@
+        this.numbers = numbers;
+    }
+
+    public static LottoTicket from(String rawNumbers) {
```
정적 메서드를 활용한 생성하고 있군요 💯

findstar commented on src/main/java/lotto/domain/LottoRank.java:
```diff
@@ -0,0 +1,53 @@
+        this.price = price;
+    }
+
+    public static LottoRank get(int matchCount, boolean isBonus) {
```
LottoRank 에서 랭크를 구하고 있네요. 좋습니다 💯

findstar reviewed (approved):
안녕하세요 타스. 제임스입니다.
지난번 코멘트 드린 피드백 반영 잘 해주셨네요. 
궁금한 사항은 언제든 편하게 질문주세요.
수고하셨습니다. 다음 단계 진행해주세요 😄

chanhalee committed:
docs(README) Todo List 작성

Todo List 작성

chanhalee committed:
feat(Budget) 구매 금액에 대한 요구사항 Test 작성 및 구현

구매 금액에 대한 요구사항 Test 작성 및 구현

chanhalee committed:
feat(Ball) 로또 번호에 대한 요구사항 test 작성 및 구현

로또 번호에 대한 요구사항 test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 정수여야 한다.

chanhalee committed:
feat(Ticket) 티켓에 대한 요구사항 Test 작성 및 구현

티켓에 대한 요구사항 Test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 중복되지 않는 6개의 수로 이루어져 있다

chanhalee committed:
refactor(Budget) validation 메서드 static 속성 제거

validation 메서드 static 속성 제거

chanhalee committed:
feat(WinningNumbers) 티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현

티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현
- 등수에 대한 Enum 클래스 Prize 작성
- Ball -> Number로 클래스 명 변경
- Number Set을 보유하는 Numbers 클래스 생성
- WinningNumbers, TicketNumbers가 Numbers를 상속 받도록 구현

chanhalee committed:
feat(PrizeCalculator) 수익률계산기 테스트 작성 및 구현

수익률계산기 테스트 작성 및 구현

chanhalee committed:
feat(LottoMachine) LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

chanhalee committed:
feat(LottoMachine) LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

chanhalee committed:
refactor MVC 패턴에 맞게 패키지 분리

chanhalee committed:
feat(LottoView) view 레이어 구현

view 레이어 구현

chanhalee committed:
feat(LottoController) Controller 구현

Controller 구현

seokhong reviewed (changes requested):
찬하님 안녕하세요.
로또 미션을 함께 진행할 리뷰어 김석홍입니다. 만나서 반갑습니다!

코멘트 남긴 내용 확인 부탁드립니다.
코멘트 관련하여 궁금하신 점은 PR에 코멘트 또는 슬랙으로 DM 보내주시면 확인하는대로 회신 드리겠습니다. 🙂

chanhalee committed:
Refactor

리뷰어의 코멘트를 반영하여 리펙터링

seokhong reviewed (approved):
찬하님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
다음 단계 미션 진행을 위해 1단계 PR은 머지하겠습니다.
질문 남겨주신 내용에 대한 답변과 함께 추가로 고민해보시면 좋을 내용 코멘트로 남겼습니다. 다음 단계 진행하실 때 참고해 보시면 좋겠습니다.
2단계 미션에서 뵙겠습니다. 🙂
